File: frontend/src/products/vision/services/callBellService.ts
import type {
  CallBellEvent,
  CallBellStaffMetrics,
  CallBellFloorMetrics,
  CallBellShiftMetrics,
  CallBellDailySummary,
  PaginatedResponse,
} from '@/types';
import { mockStore } from '@/mock';
import { simulateLatency } from '@/platform/services/api';

interface CallBellFilters {
  unit?: string;
  floor?: number;
  shift?: string;
  priority?: string;
  status?: string;
  dateFrom?: string;
  dateTo?: string;
  page?: number;
  pageSize?: number;
}

export async function getCallBellEvents(
  filters: CallBellFilters = {}
): Promise<PaginatedResponse<CallBellEvent>> {
  await simulateLatency();
  let events = [...mockStore.callBellEvents];

  if (filters.unit) events = events.filter(e => e.unit === filters.unit);
  if (filters.floor !== undefined) events = events.filter(e => e.floor === filters.floor);
  if (filters.shift) events = events.filter(e => e.shift === filters.shift);
  if (filters.priority) events = events.filter(e => e.priority === filters.priority);
  if (filters.status) events = events.filter(e => e.status === filters.status);
  if (filters.dateFrom) {
    const from = new Date(filters.dateFrom).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() >= from);
  }
  if (filters.dateTo) {
    const to = new Date(filters.dateTo).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() <= to);
  }

  const page = filters.page || 1;
  const pageSize = filters.pageSize || 20;
  const total = events.length;
  const totalPages = Math.ceil(total / pageSize);
  const start = (page - 1) * pageSize;
  const data = events.slice(start, start + pageSize);

  return { data, total, page, pageSize, totalPages };
}

export async function getActiveCallBells(): Promise<CallBellEvent[]> {
  await simulateLatency();
  return mockStore.getActiveCallBells();
}

export async function respondToCallBell(
  eventId: string,
  userId: string,
  userName: string
): Promise<CallBellEvent | undefined> {
  await simulateLatency();
  return mockStore.respondToCallBell(eventId, userId, userName);
}

export async function closeCallBell(eventId: string): Promise<CallBellEvent | undefined> {
  await simulateLatency();
  return mockStore.closeCallBell(eventId);
}

export async function getStaffMetrics(
  filters: { unit?: string; dateFrom?: string; dateTo?: string } = {}
): Promise<CallBellStaffMetrics[]> {
  await simulateLatency();
  let events = mockStore.callBellEvents.filter(
    e => e.status === 'responded' || e.status === 'closed'
  );

  if (filters.unit) events = events.filter(e => e.unit === filters.unit);
  if (filters.dateFrom) {
    const from = new Date(filters.dateFrom).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() >= from);
  }
  if (filters.dateTo) {
    const to = new Date(filters.dateTo).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() <= to);
  }

  const staffMap = new Map<string, CallBellEvent[]>();
  for (const e of events) {
    if (!e.respondedBy) continue;
    const existing = staffMap.get(e.respondedBy) || [];
    existing.push(e);
    staffMap.set(e.respondedBy, existing);
  }

  const metrics: CallBellStaffMetrics[] = [];
  for (const [staffId, staffEvents] of staffMap) {
    const times = staffEvents
      .map(e => e.responseTimeSeconds)
      .filter((t): t is number => t !== null)
      .sort((a, b) => a - b);

    if (times.length === 0) continue;

    const avg = Math.round(times.reduce((s, t) => s + t, 0) / times.length);
    const median = times[Math.floor(times.length / 2)];

    metrics.push({
      staffId,
      staffName: staffEvents[0].respondedByName || staffId,
      role: staffId.startsWith('nurse') ? 'Nurse' : 'PSW',
      totalCalls: staffEvents.length,
      avgResponseSeconds: avg,
      medianResponseSeconds: median,
      minResponseSeconds: times[0],
      maxResponseSeconds: times[times.length - 1],
      callsUnder60s: times.filter(t => t <= 60).length,
      callsUnder120s: times.filter(t => t <= 120).length,
      callsOver180s: times.filter(t => t > 180).length,
    });
  }

  metrics.sort((a, b) => a.avgResponseSeconds - b.avgResponseSeconds);
  return metrics;
}

export async function getFloorMetrics(
  filters: { dateFrom?: string; dateTo?: string } = {}
): Promise<CallBellFloorMetrics[]> {
  await simulateLatency();
  let events = mockStore.callBellEvents.filter(e => e.responseTimeSeconds !== null);

  if (filters.dateFrom) {
    const from = new Date(filters.dateFrom).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() >= from);
  }
  if (filters.dateTo) {
    const to = new Date(filters.dateTo).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() <= to);
  }

  const floorMap = new Map<string, CallBellEvent[]>();
  for (const e of events) {
    const key = `${e.floor}-${e.unit}`;
    const existing = floorMap.get(key) || [];
    existing.push(e);
    floorMap.set(key, existing);
  }

  const metrics: CallBellFloorMetrics[] = [];
  for (const [, floorEvents] of floorMap) {
    const times = floorEvents
      .map(e => e.responseTimeSeconds)
      .filter((t): t is number => t !== null)
      .sort((a, b) => a - b);

    const avg = times.length > 0
      ? Math.round(times.reduce((s, t) => s + t, 0) / times.length)
      : 0;

    const callsByPriority = { emergency: 0, urgent: 0, normal: 0 } as Record<string, number>;
    for (const e of floorEvents) callsByPriority[e.priority]++;

    const callsByOrigin = { bedside: 0, bathroom: 0, hallway: 0, pendant: 0 } as Record<string, number>;
    for (const e of floorEvents) callsByOrigin[e.origin]++;

    const hourCounts = new Array(24).fill(0);
    for (const e of floorEvents) {
      hourCounts[new Date(e.pressedAt).getHours()]++;
    }
    const peakHour = hourCounts.indexOf(Math.max(...hourCounts));

    metrics.push({
      floor: floorEvents[0].floor,
      unit: floorEvents[0].unit,
      totalCalls: floorEvents.length,
      avgResponseSeconds: avg,
      callsByPriority: callsByPriority as CallBellFloorMetrics['callsByPriority'],
      callsByOrigin: callsByOrigin as CallBellFloorMetrics['callsByOrigin'],
      peakHour,
    });
  }

  return metrics;
}

export async function getShiftMetrics(
  filters: { unit?: string; dateFrom?: string; dateTo?: string } = {}
): Promise<CallBellShiftMetrics[]> {
  await simulateLatency();
  let events = [...mockStore.callBellEvents];

  if (filters.unit) events = events.filter(e => e.unit === filters.unit);
  if (filters.dateFrom) {
    const from = new Date(filters.dateFrom).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() >= from);
  }
  if (filters.dateTo) {
    const to = new Date(filters.dateTo).getTime();
    events = events.filter(e => new Date(e.pressedAt).getTime() <= to);
  }

  // Group by shift name (aggregate across all dates/units)
  const shiftMap = new Map<string, CallBellEvent[]>();
  for (const e of events) {
    const existing = shiftMap.get(e.shift) || [];
    existing.push(e);
    shiftMap.set(e.shift, existing);
  }

  const metrics: CallBellShiftMetrics[] = [];
  for (const [shift, shiftEvents] of shiftMap) {
    const times = shiftEvents
      .map(e => e.responseTimeSeconds)
      .filter((t): t is number => t !== null)
      .sort((a, b) => a - b);

    const avg = times.length > 0
      ? Math.round(times.reduce((s, t) => s + t, 0) / times.length)
      : 0;

    metrics.push({
      shift: shift as 'Day' | 'Evening' | 'Night',
      date: '',
      unit: filters.unit || 'All',
      totalCalls: shiftEvents.length,
      avgResponseSeconds: avg,
      respondedWithin60s: times.filter(t => t <= 60).length,
      respondedWithin120s: times.filter(t => t <= 120).length,
      slowResponses: times.filter(t => t > 180).length,
    });
  }

  // Sort Day, Evening, Night
  const order = { Day: 0, Evening: 1, Night: 2 };
  metrics.sort((a, b) => order[a.shift] - order[b.shift]);
  return metrics;
}

export async function getDailySummaries(
  filters: { unit?: string } = {}
): Promise<CallBellDailySummary[]> {
  await simulateLatency();
  let events = [...mockStore.callBellEvents];
  if (filters.unit) events = events.filter(e => e.unit === filters.unit);

  const dayMap = new Map<string, CallBellEvent[]>();
  for (const e of events) {
    const date = e.pressedAt.slice(0, 10);
    const existing = dayMap.get(date) || [];
    existing.push(e);
    dayMap.set(date, existing);
  }

  const summaries: CallBellDailySummary[] = [];
  for (const [date, dayEvents] of dayMap) {
    const responded = dayEvents.filter(e => e.responseTimeSeconds !== null);
    const avg = responded.length > 0
      ? Math.round(responded.reduce((s, e) => s + e.responseTimeSeconds!, 0) / responded.length)
      : 0;

    summaries.push({ date, totalCalls: dayEvents.length, avgResponseSeconds: avg });
  }

  summaries.sort((a, b) => a.date.localeCompare(b.date));
  return summaries;
}

File: backend/src/services/call_bell_service.py
from collections import Counter, defaultdict
from datetime import datetime, timezone

from sqlalchemy import desc, select
from sqlalchemy.ext.asyncio import AsyncSession

from src.models.call_bell import CallBellEvent
from src.utils.pagination import paginate
from src.utils.time_utils import get_shift_name, seconds_between


# ---------------------------------------------------------------------------
# Query helpers
# ---------------------------------------------------------------------------


def _apply_common_filters(query, filters: dict):
    """Apply common filter predicates to a CallBellEvent query."""
    if filters.get("unit"):
        query = query.where(CallBellEvent.unit == filters["unit"])
    if filters.get("floor"):
        query = query.where(CallBellEvent.floor == filters["floor"])
    if filters.get("shift"):
        query = query.where(CallBellEvent.shift == filters["shift"])
    if filters.get("priority"):
        query = query.where(CallBellEvent.priority == filters["priority"])
    if filters.get("status"):
        query = query.where(CallBellEvent.status == filters["status"])
    if filters.get("date_from"):
        query = query.where(CallBellEvent.pressed_at >= filters["date_from"])
    if filters.get("date_to"):
        query = query.where(CallBellEvent.pressed_at <= filters["date_to"])
    return query


async def _load_filtered_events(
    db: AsyncSession, filters: dict
) -> list[CallBellEvent]:
    """Load call bell events matching filters (unpaginated, for metrics)."""
    query = select(CallBellEvent).order_by(desc(CallBellEvent.pressed_at))
    query = _apply_common_filters(query, filters)
    result = await db.execute(query)
    return list(result.scalars().all())


def _median(values: list[int | float]) -> float:
    """Compute median of a list of numbers."""
    if not values:
        return 0.0
    s = sorted(values)
    mid = len(s) // 2
    if len(s) % 2 == 0:
        return round((s[mid - 1] + s[mid]) / 2, 1)
    return float(s[mid])


# ---------------------------------------------------------------------------
# CRUD
# ---------------------------------------------------------------------------


async def get_call_bell_events(db: AsyncSession, filters: dict) -> dict:
    """Get paginated call bell events with filters.

    Supported filters: unit, floor, shift, priority, status,
    date_from, date_to, page, page_size.
    """
    query = select(CallBellEvent).order_by(desc(CallBellEvent.pressed_at))
    query = _apply_common_filters(query, filters)

    page = filters.get("page", 1)
    page_size = filters.get("page_size", 20)
    return await paginate(db, query, page=page, page_size=page_size)


async def get_active_call_bells(db: AsyncSession) -> list[CallBellEvent]:
    """Get call bells that are still active or responded (not yet closed)."""
    result = await db.execute(
        select(CallBellEvent)
        .where(CallBellEvent.status.in_(["active", "responded"]))
        .order_by(desc(CallBellEvent.pressed_at))
    )
    return list(result.scalars().all())


async def respond_to_call_bell(
    db: AsyncSession, event_id: str, user_id: str, user_name: str
) -> CallBellEvent | None:
    """Mark a call bell as responded and compute response time."""
    event = await db.get(CallBellEvent, event_id)
    if not event or event.status != "active":
        return None

    now = datetime.now(timezone.utc)
    event.status = "responded"
    event.responded_at = now
    event.responded_by = user_id
    event.responded_by_name = user_name

    if event.pressed_at:
        event.response_time_seconds = seconds_between(event.pressed_at, now)

    await db.commit()
    await db.refresh(event)
    return event


async def close_call_bell(
    db: AsyncSession, event_id: str
) -> CallBellEvent | None:
    """Mark a call bell as closed."""
    event = await db.get(CallBellEvent, event_id)
    if not event or event.status not in ("active", "responded"):
        return None

    event.status = "closed"
    event.closed_at = datetime.now(timezone.utc)
    await db.commit()
    await db.refresh(event)
    return event


# ---------------------------------------------------------------------------
# Staff metrics
# ---------------------------------------------------------------------------


async def get_staff_metrics(db: AsyncSession, filters: dict) -> list[dict]:
    """Compute per-staff response metrics.

    Groups by responded_by. Computes avg, min, max, median response times
    and counts calls under 60s, under 120s, over 180s.
    """
    events = await _load_filtered_events(db, filters)

    # Only include events that have a responder and a response time
    staff_events: dict[str, list[CallBellEvent]] = defaultdict(list)
    for e in events:
        if e.responded_by and e.response_time_seconds is not None:
            staff_events[e.responded_by].append(e)

    metrics: list[dict] = []
    for staff_id, evts in staff_events.items():
        times = [e.response_time_seconds for e in evts if e.response_time_seconds is not None]
        if not times:
            continue

        # Use the name from the first event that has one
        staff_name = next(
            (e.responded_by_name for e in evts if e.responded_by_name), staff_id
        )

        metrics.append(
            {
                "staff_id": staff_id,
                "staff_name": staff_name,
                "role": "staff",
                "total_calls": len(evts),
                "avg_response_seconds": round(sum(times) / len(times), 1),
                "median_response_seconds": _median(times),
                "min_response_seconds": float(min(times)),
                "max_response_seconds": float(max(times)),
                "calls_under_60s": sum(1 for t in times if t < 60),
                "calls_under_120s": sum(1 for t in times if t < 120),
                "calls_over_180s": sum(1 for t in times if t > 180),
            }
        )

    return metrics


# ---------------------------------------------------------------------------
# Floor metrics
# ---------------------------------------------------------------------------


async def get_floor_metrics(db: AsyncSession, filters: dict) -> list[dict]:
    """Compute per-floor/unit call bell metrics.

    Groups by (floor, unit). Computes calls by priority and origin, and
    finds the peak hour.
    """
    events = await _load_filtered_events(db, filters)

    floor_events: dict[tuple[int | None, str | None], list[CallBellEvent]] = defaultdict(list)
    for e in events:
        floor_events[(e.floor, e.unit)].append(e)

    metrics: list[dict] = []
    for (floor_val, unit_val), evts in floor_events.items():
        times = [
            e.response_time_seconds
            for e in evts
            if e.response_time_seconds is not None
        ]
        avg_response = round(sum(times) / len(times), 1) if times else 0.0

        priority_counts: dict[str, int] = dict(
            Counter(e.priority for e in evts if e.priority)
        )
        origin_counts: dict[str, int] = dict(
            Counter(e.origin for e in evts if e.origin)
        )

        # Peak hour
        hour_counts: dict[int, int] = Counter(
            e.pressed_at.hour for e in evts if e.pressed_at
        )
        peak_hour = max(hour_counts, key=hour_counts.get) if hour_counts else 0

        metrics.append(
            {
                "floor": floor_val or 0,
                "unit": unit_val or "unknown",
                "total_calls": len(evts),
                "avg_response_seconds": avg_response,
                "calls_by_priority": priority_counts,
                "calls_by_origin": origin_counts,
                "peak_hour": peak_hour,
            }
        )

    return metrics


# ---------------------------------------------------------------------------
# Shift metrics
# ---------------------------------------------------------------------------


async def get_shift_metrics(db: AsyncSession, filters: dict) -> list[dict]:
    """Compute per-shift call bell metrics.

    Groups by (shift, date, unit).
    """
    events = await _load_filtered_events(db, filters)

    shift_events: dict[tuple[str, str, str], list[CallBellEvent]] = defaultdict(list)
    for e in events:
        shift = e.shift or (
            get_shift_name(e.pressed_at.hour) if e.pressed_at else "Unknown"
        )
        date_str = e.pressed_at.strftime("%Y-%m-%d") if e.pressed_at else "unknown"
        unit_val = e.unit or "unknown"
        shift_events[(shift, date_str, unit_val)].append(e)

    metrics: list[dict] = []
    for (shift, date_str, unit_val), evts in shift_events.items():
        times = [
            e.response_time_seconds
            for e in evts
            if e.response_time_seconds is not None
        ]
        avg_response = round(sum(times) / len(times), 1) if times else 0.0

        metrics.append(
            {
                "shift": shift,
                "date": date_str,
                "unit": unit_val,
                "total_calls": len(evts),
                "avg_response_seconds": avg_response,
                "responded_within_60s": sum(1 for t in times if t <= 60),
                "responded_within_120s": sum(1 for t in times if t <= 120),
                "slow_responses": sum(1 for t in times if t > 180),
            }
        )

    return metrics


# ---------------------------------------------------------------------------
# Daily summaries
# ---------------------------------------------------------------------------


async def get_daily_summaries(
    db: AsyncSession, unit: str | None = None
) -> list[dict]:
    """Compute daily call bell summaries grouped by date."""
    query = select(CallBellEvent).order_by(desc(CallBellEvent.pressed_at))
    if unit:
        query = query.where(CallBellEvent.unit == unit)

    result = await db.execute(query)
    events = list(result.scalars().all())

    daily: dict[str, list[CallBellEvent]] = defaultdict(list)
    for e in events:
        if e.pressed_at:
            date_str = e.pressed_at.strftime("%Y-%m-%d")
            daily[date_str].append(e)

    summaries: list[dict] = []
    for date_str in sorted(daily.keys(), reverse=True):
        evts = daily[date_str]
        times = [
            e.response_time_seconds
            for e in evts
            if e.response_time_seconds is not None
        ]
        avg_response = round(sum(times) / len(times), 1) if times else 0.0

        summaries.append(
            {
                "date": date_str,
                "total_calls": len(evts),
                "avg_response_seconds": avg_response,
            }
        )

    return summaries
